### PR TITLE
AIP-72: Adding support to Set an XCom from Task SDK

### LINKED ISSUE #44350: Add Endpoint to Push XCom from Task SDK
Similar to the XCom get endpoint in the Task Execution API (added in [this PR](https://github.com/apache/airflow/pull/44101)), we should add a way to push XCom from a Task to API server (via Supervisor)

### PR DESCRIPTION
**Pull request by amoghrajesh**
closes: https://github.com/apache/airflow/issues/44350

Adding support to be able to put/create xcoms from the task sdk side.

#### Supervisor
The supervisor would receive a request from the task runner to CREATE a xcoms.
The handle_request method handles this part and is responsible to send a request to the API client to send it to the tas execution interface.
In this case, SetXCom is the body that will be passed down to the api client.

#### API Client
The client receives a request from the supervisor to CREATE a xcom. The client then sends it to the task execution interface over HTTP. It send a POST request with the received body. The key isn't stripped off but is sent in the payload even though the server doesn't serialise it.
The client receives a response {"message": "XCom successfully set"} if all's good.
It passes this response back down to the supervisor which then writes it in the buffer.

#### Testing
Since end to end isn't possible, the test case https://github.com/apache/airflow/compare/main...astronomer:airflow:AIP72-set-xcom-tasksdk?expand=1#diff-f55b9ba6be46de22b2838085bf1e0aa4193797c3ff7d0cdc4e06a626c96df1baR693-R713 checks for supervisor side of things, if supervisor processes "request" of correct format correctly. Checks for call args, responses, even buffer.

The test https://github.com/apache/airflow/compare/main...astronomer:airflow:AIP72-set-xcom-tasksdk?expand=1#diff-048659c5ed2897cd0ed24a340885d5950c964ea451ccc359e1ab592bf8a59f0dR138-R165 checks if the client can handle requests of type SetXCom properly and respond accordingly.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: task_sdk/src/airflow/sdk/api/client.py [hunk 1]
```diff
@@ -169,6 +169,14 @@ def get(self, dag_id: str, run_id: str, task_id: str, key: str, map_index: int =
         resp = self.client.get(f"xcoms/{dag_id}/{run_id}/{task_id}/{key}", params={"map_index": map_index})
         return XComResponse.model_validate_json(resp.read())
 
+    def set(self, dag_id: str, run_id: str, task_id: str, key: str, value, map_index: int | None = None):
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/api/client.py]:
```suggestion
    def set(self, dag_id: str, run_id: str, task_id: str, key: str, value, map_index: int | None = None) -> dict[str, bool]:
```

File: task_sdk/tests/api/test_client.py [hunk 2]
```diff
@@ -133,3 +133,33 @@ def handle_request(request: httpx.Request) -> httpx.Response:
...
+
+
+class TestXCOMOperations:
+    """
+    Test that the XComOperations class works as expected. While the operations are simple, it
+    still catches the basic functionality of the client for xcoms including endpoint and
+    response parsing.
+    """
+
+    # TODO: Add tests for get xcom in a follow up
```
Comment (kaxil) [inline task_sdk/tests/api/test_client.py]: You can add it in this one too if you want.. should be a simple test

File: task_sdk/tests/api/test_client.py [hunk 3]
```diff
@@ -133,3 +133,33 @@ def handle_request(request: httpx.Request) -> httpx.Response:
...
+
+    def test_xcom_set_success(self):
+        # Simulate a successful response from the server when setting an xcom
+        def handle_request(request: httpx.Request) -> httpx.Response:
+            if request.url.path == "/xcoms/dag_id/run_id/task_id/key":
+                return httpx.Response(
+                    status_code=200,
```
Comment (kaxil) [inline task_sdk/tests/api/test_client.py]: Server sends 201

File: task_sdk/tests/execution_time/test_supervisor.py [hunk 4]
```diff
@@ -689,6 +690,27 @@ def watched_subprocess(self, mocker):
...
+                "xcoms.set",
+                (
+                    "test_dag",
+                    "test_run",
+                    "test_task",
+                    "test_key",
+                    '{"key": "test_key", "value": {"key2": "value2"}}',
+                    None,
+                ),
+                {"message": "XCom successfully set"},
```
Comment (kaxil) [inline task_sdk/tests/execution_time/test_supervisor.py]: This needs changing, now that you use `{"ok": True}`

File: task_sdk/tests/api/test_client.py [hunk 5]
```diff
@@ -133,3 +133,33 @@ def handle_request(request: httpx.Request) -> httpx.Response:
...
+                )
+            return httpx.Response(status_code=400, json={"detail": "Bad Request"})
+
+        client = make_client(transport=httpx.MockTransport(handle_request))
+        result = client.xcoms.set(
+            dag_id="dag_id",
+            run_id="run_id",
+            task_id="task_id",
+            key="key",
+            value={"key": "test_key", "value": {"key2": "value2"}},
```
Comment (kaxil) [inline task_sdk/tests/api/test_client.py]: Can you use parameterize to verify that we can handle more types: list, str etc

File: task_sdk/src/airflow/sdk/api/client.py (hunk 1 above)
Comment (amoghrajesh) [reply task_sdk/src/airflow/sdk/api/client.py]: Handled

File: task_sdk/tests/api/test_client.py (hunk 2 above)
Comment (amoghrajesh) [reply task_sdk/tests/api/test_client.py]: Sure, adding in this PR itself

File: task_sdk/tests/api/test_client.py (hunk 3 above)
Comment (amoghrajesh) [reply task_sdk/tests/api/test_client.py]: Right. And also `{"message": "XCom successfully set"},`

File: task_sdk/tests/execution_time/test_supervisor.py (hunk 4 above)
Comment (amoghrajesh) [reply task_sdk/tests/execution_time/test_supervisor.py]: Changed

File: task_sdk/tests/api/test_client.py (hunk 5 above)
Comment (amoghrajesh) [reply task_sdk/tests/api/test_client.py]: Handled it in cc62d5483800f40d5297e55ceac56cb284494911

File: task_sdk/src/airflow/sdk/api/client.py [hunk 6]
```diff
@@ -169,6 +169,19 @@ def get(self, dag_id: str, run_id: str, task_id: str, key: str, map_index: int =
...
+        self, dag_id: str, run_id: str, task_id: str, key: str, value, map_index: int | None = None
+    ) -> dict[str, bool]:
+        """Set a XCom value via the API server."""
+        params = {}
+        if map_index:
+            params = {"map_index": map_index}
+        self.client.post(f"xcoms/{dag_id}/{run_id}/{task_id}/{key}", params=params, json=value)
+        # Any error from the server will anyway be propagated down to the supervisor,
+        # so we choose to send a generic response to the supervisor over the server response to
+        # decouple from the server response string
```
Comment (amoghrajesh) [inline task_sdk/src/airflow/sdk/api/client.py:196]: Same as https://github.com/apache/airflow/pull/44562#discussion_r1868014142

File: task_sdk/tests/api/test_client.py (hunk 2 above)
Comment (amoghrajesh) [reply task_sdk/tests/api/test_client.py]: Handled in 12fd0c8cdc1fb7675fa2d925fc2a008026dc7e45

File: task_sdk/src/airflow/sdk/api/client.py [hunk 7]
```diff
@@ -169,6 +169,19 @@ def get(self, dag_id: str, run_id: str, task_id: str, key: str, map_index: int =
...
         return XComResponse.model_validate_json(resp.read())
 
+    def set(
+        self, dag_id: str, run_id: str, task_id: str, key: str, value, map_index: int | None = None
+    ) -> dict[str, bool]:
+        """Set a XCom value via the API server."""
+        params = {}
+        if map_index:
+            params = {"map_index": map_index}
+        self.client.post(f"xcoms/{dag_id}/{run_id}/{task_id}/{key}", params=params, json=value)
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/api/client.py:193]: I'm not sure map_index should be a query param. I think in Airflow API when a task is mapped we add an extra `f"/{map_index}` part to the URL. Does that sound right/familiar to you @uranusjr @jedcunningham or am I making stuff up again?

File: task_sdk/tests/api/test_client.py
Comment (ashb) [inline task_sdk/tests/api/test_client.py:1]: Lets add a test that ensures that when a map_index is passed the right url/query param is passed too.

File: task_sdk/src/airflow/sdk/api/client.py (hunk 7 above)
Comment (amoghrajesh) [reply task_sdk/src/airflow/sdk/api/client.py:193]: Tried with map_index as a query param and it works:
image

Even its usage in https://github.com/apache/airflow/blob/main/airflow/api_fastapi/core_api/routes/public/xcom.py#L57 hints that it is a query param.

Even tests for that matter:
```
        response = test_client.get(
            "/public/dags/~/dagRuns/~/taskInstances/~/xcomEntries",
            params={"map_index": map_index} if map_index is not None else None,
        )
```

File: task_sdk/tests/api/test_client.py
Comment (amoghrajesh) [reply task_sdk/tests/api/test_client.py:1]: Sure, can do

Comment (amoghrajesh) [reply task_sdk/tests/api/test_client.py:1]: Handled it in 7195e1c73b36a773150ba37710da0031c7407aa1

File: task_sdk/src/airflow/sdk/api/client.py (hunk 7 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/api/client.py:193]: https://github.com/apache/airflow/blob/v2-10-stable/airflow/api_connexion/openapi/v1.yaml#L1785C1-L1785C81 is what I was basing it off

```
  /dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}/{map_index}/tries:
```

Comment (ashb) [reply task_sdk/src/airflow/sdk/api/client.py:193]: But okay, this is consistent, we can revisit both of these later if we want to.

Comment (amoghrajesh) [reply task_sdk/src/airflow/sdk/api/client.py:193]: Should I leave a comment on this one and we can check again later?

Comment (amoghrajesh) [reply task_sdk/src/airflow/sdk/api/client.py:193]: I just added a TODO in the code for this part, we can revisit it always once we start testing things

File: task_sdk/tests/api/test_client.py [hunk 8]
```diff
@@ -148,3 +148,110 @@ def handle_request(request: httpx.Request) -> httpx.Response:
...
+            pytest.param({"key1": "value1"}, id="dict-value"),
+            pytest.param(["value1", "value2"], id="list-value"),
+            pytest.param({"key": "test_key", "value": {"key2": "value2"}}, id="nested-dict-value"),
+        ],
+    )
+    def test_xcom_set_success(self, values):
+        # Simulate a successful response from the server when setting an xcom
+        def handle_request(request: httpx.Request) -> httpx.Response:
+            if request.url.path == "/xcoms/dag_id/run_id/task_id/key":
+                return httpx.Response(
```
Comment (ashb) [inline task_sdk/tests/api/test_client.py:222]: Lets add a test/assert of something about the request body please

**Review by ashb — APPROVED** [review]
One small change (lets assert something about the request body) then this one is good to merge

File: task_sdk/tests/api/test_client.py (hunk 8 above)
Comment (amoghrajesh) [reply task_sdk/tests/api/test_client.py:222]: Yeah makes sense. Added it in 420ac000ede7f2f9580bf2622eebcd17b2d0a79f
 Does it seem ok/

**Pull request merged**